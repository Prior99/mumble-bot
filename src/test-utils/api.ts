import * as SuperTest from "supertest";
import { RestApi } from "../server";

export const api = () => SuperTest(tsdi.get(RestApi).app);
