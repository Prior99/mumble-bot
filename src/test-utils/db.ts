import { Connection } from "typeorm";
import { DatabaseFactory, MumbleFactory, AudioInput } from "../server";

export async function startDb() {
    const databaseFactory = tsdi.get(DatabaseFactory);
    await databaseFactory.connect();
    const db = tsdi.get(Connection);
    await db.query(`
        DROP SCHEMA public CASCADE;
        CREATE SCHEMA public;
        CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
    `);
    await db.runMigrations();
    await tsdi.get(MumbleFactory).connect();
    await tsdi.get(AudioInput).initialize();
}

export async function stopDb() {
    const db = tsdi.get(Connection);
    // This needs to be performed in order to flush all active queries.
    // There might be queries ongoing which, on termination will fail all tests.
    // By executing this dummy query it is ensured that the database driver waits for
    // all queries before closing.
    await db.query("SELECT 1");
    await db.close();
}
