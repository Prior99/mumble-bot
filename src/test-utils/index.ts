export * from "./api";
export * from "./mumble-link";
export * from "./token";
export * from "./user";
export * from "./sound";
export * from "./tag";
export * from "./playlist";
export * from "./db";
