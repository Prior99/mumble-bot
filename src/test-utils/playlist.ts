import { Sound, User, Playlist, PlaylistEntry } from "../common";
import { Connection } from "typeorm";

const defaultPlaylist = {
    name: "A Playlist",
};

export async function createPlaylist(creator: User, ...sounds: Sound[]) {
    const playlist = await tsdi.get(Connection).getRepository(Playlist).save(Object.assign(new Playlist(), {
        ...defaultPlaylist,
        creator,
    }));
    await Promise.all(sounds.map(async (sound, position) => {
        await tsdi.get(Connection).getRepository(PlaylistEntry).save({
            sound, position, playlist,
        });
    }));
    return await tsdi.get(Connection).getRepository(Playlist).findOne({
        where: { id: playlist.id },
        relations: ["creator", "entries", "entries.sound"],
    });
}
