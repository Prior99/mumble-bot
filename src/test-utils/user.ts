import { Connection } from "typeorm";
import { api } from "./api";
import { User } from "../common";

const defaultUser = {
    name: "someone",
    email: "some@example.com",
    password: "some secure password",
};

export async function createUser(data?: User, enable = true, admin = false) {
    const response = await api().post("/user").send({ ...defaultUser, ...data });
    const user = response.body.data;
    if (enable) {
        await tsdi.get(Connection).getRepository(User).update(user.id, { enabled: true });
        user.enabled = true;
    }
    if (admin) {
        await tsdi.get(Connection).getRepository(User).update(user.id, { admin: true });
        user.admin = true;
    }
    return user;
}
