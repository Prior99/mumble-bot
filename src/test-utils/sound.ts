import { Sound, User } from "../common";
import { Connection } from "typeorm";
import { createUser } from "./user";
import { api } from "./api";

const defaultSound = {
    description: "Some sound",
    used: 17,
    source: "recording",
    duration: 13,
};

export async function createSound(sound?: Sound) {
    return await tsdi.get(Connection).getRepository(Sound).save({ ...defaultSound, ...sound } as Sound);
}

export async function createSoundWithCreatorAndSpeaker(data?: Sound) {
    const user = await createUser({ name: "Speaker", email: "speaker@example.com" } as User);
    const creator = await createUser({ name: "Creator", email: "creator@example.com" } as User);
    const sound = await createSound({ ...data, user, creator });
    return { sound, creator, speaker: user };
}

export async function getSound(id: string, token: string) {
    const response = await api().get(`/sound/${id}`)
        .set("authorization", `Bearer ${token}`);
    return response.body.data;
}

export async function rateSound(id: string, token: string, stars: number) {
    return await api().post(`/sound/${id}/ratings`)
        .set("authorization", `Bearer ${token}`)
        .send({ stars });
}
