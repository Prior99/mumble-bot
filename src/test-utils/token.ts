import { api } from "./api";
import { createUserManually, createUser } from "./user";
import { User } from "../common";

export async function createUserWithToken(data?: User) {
    const user = await createUser(data);
    const response = await api().post("/token")
        .send({
            password: "some secure password",
            email: user.email,
        });
    const token = response.body.data;
    return { token, user };
}

export async function createUserWithTokenManually(data?: User) {
    const user = await createUserManually(data);
    const response = await api().post("/token")
        .send({
            password: "some secure password",
            email: user.email,
        });
    const token = response.body.data;
    return { token, user };
}
