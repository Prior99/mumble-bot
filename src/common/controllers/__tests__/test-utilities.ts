import { api, createUserWithToken, startDb, stopDb } from "../../../test-utils";

describe("utilities controller", () => {
    beforeEach(startDb);
    afterEach(stopDb);

    describe("GET /channel-tree", () => {
        it("responds 401 without a valid token", async () => {
            const response = await api().get("/channel-tree");
            expect(response.body).toEqual({ message: "Unauthorized." });
            expect(response.status).toBe(401);
        });

        it("fetches a tree of the server's channels", async () => {
            const { token } = await createUserWithToken();
            const response = await api().get("/channel-tree")
                .set("authorization", `Bearer ${token.id}`);
            expect(response.body).toMatchSnapshot();
            expect(response.status).toBe(200);
        });
    });

    describe("GET /mumble-users", () => {
        it("responds 401 without a valid token", async () => {
            const response = await api().get("/mumble-users");
            expect(response.body).toEqual({ message: "Unauthorized." });
            expect(response.status).toBe(401);
        });

        it("fetches a list of connected mumble users", async () => {
            const { token } = await createUserWithToken();
            const response = await api().get("/mumble-users")
                .set("authorization", `Bearer ${token.id}`);
            expect(response.body).toEqual({
                data: [
                    {
                        id: 4,
                        name: "Stranger one",
                        session: 84,
                    }, {
                        id: 7,
                        name: "Stranger two",
                        session: 171,
                    }, {
                        id: 23,
                        name: "Stranger three",
                        session: 2134,
                    },
                ],
            });
            expect(response.status).toBe(200);
        });
    });
});
