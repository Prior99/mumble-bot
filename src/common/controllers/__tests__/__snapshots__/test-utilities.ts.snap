// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utilities controller GET /channel-tree fetches a tree of the server's channels 1`] = `
Object {
  "data": Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "Subchild 1",
            "position": 0,
            "users": Array [
              Object {
                "id": 4,
                "name": "Stranger one",
                "selfDeaf": false,
                "selfMute": false,
                "session": 84,
              },
              Object {
                "id": 7,
                "name": "Stranger two",
                "selfDeaf": false,
                "selfMute": false,
                "session": 171,
              },
            ],
          },
        ],
        "name": "Child 1",
        "position": 0,
        "users": Array [],
      },
      Object {
        "children": Array [],
        "name": "Child 2",
        "position": 1,
        "users": Array [
          Object {
            "id": 23,
            "name": "Stranger three",
            "selfDeaf": false,
            "selfMute": false,
            "session": 2134,
          },
        ],
      },
    ],
    "name": "Root",
    "position": 0,
    "users": Array [
      Object {
        "name": "Unregistered Stranger",
        "selfDeaf": false,
        "selfMute": false,
        "session": 2134,
      },
    ],
  },
}
`;
